accepted/2021/statics-in-interfaces/README.md

```diff
@@ -6,8 +6,1267 @@
+    }
+
+    public interface ISubtractable<TSelf, TOther, TResult>
+        where TSelf : ISubtractable<TSelf, TOther, TResult>
+    {
+        static abstract TResult operator -(TSelf lhs, TOther rhs);
+
+        static abstract TResult checked operator -(TSelf lhs, TOther rhs);
+    }
+
+    public interface IIncrementable<TSelf>
+        where TSelf : IIncrementable<TSelf>
+    {
+        // We don't inherit IAddable as TOther isn't well-defined for incrementable types
+        // Incrementing twice is not necessarily the same as self + 2 and not all languages
+        // use incrementing for numbers, they are sometimes used by iterators for example
+
+        // We can't expose TSelf-- as some types only support one of the operations
+
+        // We don't support TResult as C# requires TSelf to be the return type
+
+        static abstract TSelf operator ++(TSelf value);
+
+        static abstract TSelf checked operator ++(TSelf value);
+    }
+
+    public interface IDecrementable<TSelf>
+        where TSelf : IDecrementable<TSelf>
+    {
+        static abstract TSelf operator --(TSelf value);
+
+        static abstract TSelf checked operator --(TSelf value);
+    }
+
+    public interface IMultipliable<TSelf, TOther, TResult>
+        where TSelf : IMultipliable<TSelf, TOther, TResult>
+    {
+        // A multiplicative identity is not exposed here for the same reasons as IAddable<TSelf, TOther>
+
+        // We can't assume TOther * TSelf is valid as not everything is commutative
+
+        // We can't expose TSelf / TOther as some types, such as matrices, don't support dividing with respect to TOther
+
+        // We can't inherit from IAddable<TSelf, TOther> as some types, such as matrices, support Matrix * scalar but not Matrix + scalar
+
+        static abstract TResult operator *(TSelf lhs, TOther rhs);
+
+        static abstract TResult checked operator *(TSelf lhs, TOther rhs);
+    }
+
+    public interface IDivisable<TSelf, TOther, TResult>
+        where TSelf : IDivisable<TSelf, TOther, TResult>
+    {
+        static abstract TResult operator /(TSelf lhs, TOther rhs);
+
+        static abstract TResult checked operator /(TSelf lhs, TOther rhs);
+    }
+
+    public interface IRemainder<TSelf, TOther, TResult>
+        where TSelf : IRemainder<TSelf, TOther, TResult>
+    {
+        // The ECMA name is op_Modulus and so the behavior isn't actually "remainder" with regards to negative values
+
+        // Likewise the name IRemainder doesn't fit with the other names, so a better one is needed
+
+        static abstract TResult operator %(TSelf lhs, TOther rhs);
+
+        static abstract TResult checked operator %(TSelf lhs, TOther rhs);
+    }
+
+    public interface INegatable<TSelf, TResult>
+        where TSelf : INegatable<TSelf, TResult>
+    {
+        // Should unary plus be on its own type?
+
+        static abstract TResult operator +(TSelf value);
+
+        static abstract TResult checked operator +(TSelf value);
```

Comment:
> I'm wondering why this operator+ needs to exist.

For parity with `-`. There are libraries that use `+` for correctness, alignment, and readability.

Unfortunately, its a bit late for the language to just implicitly provide it :(

Reply:
> There are libraries that use + for correctness, alignment, and readability.

I understand "readability".  Can you elaborate on how `+value` vs `value` impacts correctness?  And I'm not sure what you mean here by "alignment"?

Reply:
It allows alignment of values without fighting the formatter:
```csharp
const float x = 1.234;
const float y = -1.234;
const float z = +1.234;
```

As for correctness, once you start getting into heavy math code, you'll often end up writing series where every step involves a constant with an alternating positive/negative sign.

For example, `Sin` and `Cos` are:
![image](https://user-images.githubusercontent.com/10487869/115889205-961b5780-a408-11eb-8acc-2d133ccf4447.png)

In which case you will likely have a lookup table, where the entries are different constants that have alternating `+` vs `-` and where explicitly annotating the `+` helps with show "this constant was meant to be positive"

Reply:
Ok, so not correctness in that it's actually impacting functionality.  You just meant having + allows someone to write code that looks and reads better and thus may help them to avoid bugs.

I'm not convinced that's worth requiring every type to have another two interface methods that are always going to just return the same value.  But open for discussion I guess.